### ISSUE TITLE
Optimization level -O2 causing displayio issues

### ISSUE THREAD

**Issue by DavePutz** [adafruit/circuitpython#3250]
See issue #3247. The problem was reported on a Clue running the demo spirit level script. According to @dhalbert this kind of behavior has typically meant that something that should be volatile is not, or that there's some uninitialized variable.

**Comment by DavePutz** [comment]
So, this issue has been narrowed down to the code in spim_xfer() in ports/nrf/nrfx/drivers/src/nrfx_spim.c
It can be resolved by un-commenting the line
`#define NRFX_SPIM3_NRF52840_ANOMALY_198_WORKAROUND_ENABLED 1`
in nrfx_config.h. However, there is a comment that says un-commenting that line breaks BLE.
 @dhalbert  - do you know if that comment is correct? If it is, what options do we have for building with -O2? I did find that
I could surround spim_xfer() with #pragmas to set -Os optimization for just that function, but the use of those in production is
discouraged...

**Comment by dhalbert** [comment]
Yes, that comment is correct. The other workaround for the SPIM3 hardware bug is to give SPIM3 its own reserved 8kB block of RAM. I thought I had done that, but I made a mistake in the RAM layout, and need to fix #3147. When I do that, we can try `-O2` again.

**Comment by dhalbert** [comment]
Addendum: I wrote that comment, and also pointed to the explanation:
```
// NOTE: THIS WORKAROUND CAUSES BLE CODE TO CRASH.
// It doesn't work with the SoftDevice.
// See https://devzone.nordicsemi.com/f/nordic-q-a/33982/sdk-15-software-crash-during-spi-session
// Turn on nrfx supported workarounds for errata in Rev1 of nRF52840
#ifdef NRF52840_XXAA
//    #define NRFX_SPIM3_NRF52840_ANOMALY_198_WORKAROUND_ENABLED 1
#endif
```

**Comment by DavePutz** [comment]
Closing, fixed by PR #3281. Tested OK with a Clue and Circuitplayground Bluefruit with attached Gizmo display.

**Comment by dhalbert** [comment]
@DavePutz Thanks for testing!
